pgmgr/locking.go

```diff
@@ -0,0 +1,85 @@
+package pgmgr
+
+import (
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// LockConfig encapsulates a few settings related to controlling behavior
+// related to locks when applying migrations. The default values should be
+// appropriate for almost all use cases, though for certain workloads you may
+// need to customize them. To do so within a migration, just use normal
+// postgres SET statements:
+//    SET statement_timeout TO 10000;
+//    SET lock_timeout TO 1000;
+// Or, add a lock-config key to your pgmgr.json.
+type LockConfig struct {
+	// StatementTimeout is how long each statement in the migration has to
+	// complete, in milliseconds.
+	StatementTimeout int
+
+	// LockTimeout is how long each statement will wait to acquire its needed
+	// lock, in milliseconds. Must be less than StatementTimeout.
+	LockTimeout int
+
+	// MaxRetries is how many times pgmgr will re-try your migration when the
+	// above timeouts error out. It will not retry for other failures. Set to
+	// -1 if you would like to disable retries.
+	MaxRetries int
+
+	// RetryDelay is how many seconds pgmgr will sleep between retries
+	RetryDelay int
+}
+
+// applyDefaults overwrites 0 values to default values
+func (l *LockConfig) applyDefaults() *LockConfig {
+	if l.StatementTimeout == 0 {
+		l.StatementTimeout = 1000
+	}
+
+	if l.LockTimeout == 0 {
+		l.LockTimeout = 200
+	}
+
+	if l.MaxRetries == 0 {
+		l.MaxRetries = 10
+	}
+
+	if l.RetryDelay == 0 {
+		l.RetryDelay = 5
+	}
+
+	return l
+}
+
+const lockingErrorPattern = "ERROR:.+(canceling statement due to lock timeout|canceling statement due to statement timeout|could not obtain lock on relation)"
```

Comment:
`ERROR:.+(canceling statement due to (lock|statement) timeout|could not obtain lock on relation)`